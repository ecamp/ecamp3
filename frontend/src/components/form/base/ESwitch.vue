<template>
  <ValidationProvider
    v-slot="{ errors: veeErrors }"
    :name="name"
    :vid="veeId"
    :rules="veeRules"
    :required="required"
  >
    <v-switch
      inset
      v-bind="$attrs"
      :hide-details="hideDetails"
      :error-messages="veeErrors.concat(errorMessages)"
      :label="label || name"
      :class="[inputClass]"
      :input-value="value"
      @change="$emit('input', $event)"
      v-on="$listeners"
    >
      <!-- passing through all slots -->
      <slot v-for="(_, name) in $slots" :slot="name" :name="name" />
      <template v-for="(_, name) in $scopedSlots" :slot="name" slot-scope="slotData">
        <slot :name="name" v-bind="slotData" />
      </template>
    </v-switch>
  </ValidationProvider>
</template>

<script>
import { ValidationProvider } from 'vee-validate'
import { formComponentPropsMixin } from '@/mixins/formComponentPropsMixin.js'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'

export default {
  name: 'ESwitch',
  components: { ValidationProvider },
  mixins: [formComponentPropsMixin, formComponentMixin],
  props: {
    value: { type: Boolean, required: false },
  },
}
</script>
