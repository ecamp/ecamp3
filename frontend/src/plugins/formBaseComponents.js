// Global registration of all Vue components in folder components/form/base

import ECheckbox from '@/components/form/base/ECheckbox.vue'
import EColorPicker from '@/components/form/base/EColorPicker.vue'
import EDatePicker from '@/components/form/base/EDatePicker.vue'
import EForm from '@/components/form/base/EForm.vue'
import ERichtext from '@/components/form/base/ERichtext.vue'
import ESelect from '@/components/form/base/ESelect.vue'
import ESwitch from '@/components/form/base/ESwitch.vue'
import ETextarea from '@/components/form/base/ETextarea.vue'
import ETextField from '@/components/form/base/ETextField.vue'
import ETimePicker from '@/components/form/base/ETimePicker.vue'

class FormBaseComponentsPlugin {
  install(Vue) {
    Vue.component('ECheckbox', ECheckbox)
    Vue.component('EColorPicker', EColorPicker)
    Vue.component('EDatePicker', EDatePicker)
    Vue.component('EForm', EForm)
    Vue.component('ERichtext', ERichtext)
    Vue.component('ESelect', ESelect)
    Vue.component('ESwitch', ESwitch)
    Vue.component('ETextarea', ETextarea)
    Vue.component('ETextField', ETextField)
    Vue.component('ETimePicker', ETimePicker)
  }
}

export default new FormBaseComponentsPlugin()
