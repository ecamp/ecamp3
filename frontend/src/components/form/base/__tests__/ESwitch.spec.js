import Vue from 'vue'
import Vuetify from 'vuetify'

import formBaseComponents from '@/plugins/formBaseComponents'

import { mount as mountComponent } from '@vue/test-utils'
import ESwitch from '@/components/form/base/ESwitch.vue'
import { touch } from '@/test/util'

Vue.use(Vuetify)
Vue.use(formBaseComponents)

describe('An ESwitch', () => {
  let vuetify

  const mount = (options) => {
    const app = Vue.component('App', {
      components: { ESwitch },
      data: function () {
        return {
          data: null,
        }
      },
      template: `
        <div data-app>
          <e-switch v-model="data"/>
        </div>
      `,
    })
    return mountComponent(app, { vuetify, attachTo: document.body, ...options })
  }
  beforeEach(() => {
    vuetify = new Vuetify()
  })
  test('looks like a switch', async () => {
    const wrapper = mount()
    await wrapper.setData({ data: false })
    expect(wrapper).toMatchSnapshot('unchecked')

    await wrapper.setData({ data: true })
    expect(wrapper).toMatchSnapshot('checked')
  })

  test('is checked when initialized checked', () => {
    const wrapper = mount({
      data: function () {
        return {
          data: true,
        }
      },
    })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('true')
  })

  test('updates switch when vModel changes', async () => {
    const wrapper = mount()
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('false')

    await wrapper.setData({ data: true })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('true')

    await wrapper.setData({ data: false })
    expect(
      wrapper.find('input[type=checkbox]').element.getAttribute('aria-checked')
    ).toBe('false')
  })

  test('updates vModel when user toggles with click', async () => {
    const wrapper = mount()
    const input = wrapper.find('input')

    await input.trigger('click')
    expect(wrapper.vm.data).toBe(true)

    await input.trigger('click')
    expect(wrapper.vm.data).toBe(false)
  })

  test('updates vModel when user toggles with touch swipe', async () => {
    const wrapper = mount()

    touch(wrapper.find('.v-input--selection-controls__ripple')).start(0, 0).end(20, 0)
    expect(wrapper.vm.data).toBe(true)

    jest.resetAllMocks()
    touch(wrapper.find('.v-input--selection-controls__ripple')).start(0, 0).end(-20, 0)
    expect(wrapper.vm.data).toBe(false)
  })

  test('updates vModel when user toggles with keys', async () => {
    const wrapper = mount()
    const input = wrapper.find('input')

    input.trigger('keydown.right')
    expect(wrapper.vm.data).toBe(true)

    input.trigger('keydown.right')
    expect(wrapper.vm.data).toBe(true)

    jest.resetAllMocks()
    input.trigger('keydown.left')
    expect(wrapper.vm.data).toBe(false)
  })
})
