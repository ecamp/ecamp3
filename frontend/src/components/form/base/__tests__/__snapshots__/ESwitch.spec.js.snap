// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`An ESwitch > looks like a switch > checked 1`] = `
<div data-app>
     <span>
          <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-input--selection-controls v-input--switch v-input--switch--inset primary--text">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-input--selection-controls__input">
                              <input aria-checked="true"
                                     aria-disabled="false"
                                     id
                                     role="switch"
                                     type="checkbox"
                                     value>
                              <div class="v-input--selection-controls__ripple primary--text"></div>
                              <div class="v-input--switch__track theme--light primary--text"></div>
                              <div class="v-input--switch__thumb theme--light primary--text"><transition-stub mode="out-in"
                                                    name="fab-transition"></transition-stub></div>
                         </div>
                    </div>
               </div>
          </div>
     </span>
</div>
`;

exports[`An ESwitch > looks like a switch > unchecked 1`] = `
<div data-app>
     <span>
          <div class="v-input v-input--hide-details theme--light v-input--selection-controls v-input--switch v-input--switch--inset">
               <div class="v-input__control">
                    <div class="v-input__slot">
                         <div class="v-input--selection-controls__input">
                              <input aria-checked="false"
                                     aria-disabled="false"
                                     id
                                     role="switch"
                                     type="checkbox"
                                     value>
                              <div class="v-input--selection-controls__ripple"></div>
                              <div class="v-input--switch__track theme--light"></div>
                              <div class="v-input--switch__thumb theme--light"><transition-stub mode="out-in"
                                                    name="fab-transition"></transition-stub></div>
                         </div>
                    </div>
               </div>
          </div>
     </span>
</div>
`;
